### ISSUE TITLE
API change: make `__init__` of `Optimizer` contain the `Domain` and `DataGenerator`

### ISSUE THREAD

**Issue by mpvanderschelling** [bessagroup/f3dasm_optimize#12]
Optimization needs 2 things: the parameters space and the loss function.

## Current state
Now, the `update_step` requires the `DataGeneration`. It would be cleaner if this data generator is stored as a `self.data_generator` argument during creation of the `Optimizer`

## Requested change
These two arguments should be required during *creation* of the object. 
Also then the `_construct_model` method can collapse into the `__init__` or `__post_init__` if we are using dataclasses again

## API for Optimization

There are three ways:
- Using built-in defaults by importing a function

```
from f3dasm_optimize import cmaes
experiment_data.optimize(optimizer=cmaes(population=30, seed=123), data_generator = ..., ...)
```

- Using built-in default by string argument
```
experiment_data.optimize(optimizer='cmaes', data_generator = ..., hyperparameters={'population': 30, 'seed':123})
```

- Custom optimizer
```
class MyOptimizer(Optimizer):
    def update_step(self):
           data_generator = self.datagenerator
           data = self.data
           return ExperimentData # new samples

experiment_data.optimize(optimizer=MyOptimizer, data_generator= ... , hyperparameters({'arg1': arg1, 'arg2': arg2})
